tools/trustee-cli/src/run.rs

```diff
@@ -180,56 +194,31 @@ fn get_key_paths(base_dir: &Path, key_name: &str) -> (PathBuf, PathBuf) {
     (private_path, public_path)
 }
 
-/// Ensure a HTTPS certificate exists in the given home directory.
-///
-/// Create a self-signed certificate if it doesn't exist.
-///
-/// Returns the path to the certificate.
-fn ensure_https_cert(home_dir: &Path, private_path: &Path) -> Result<PathBuf> {
-    let certificate_path = home_dir.join("https_cert.pem");
-
-    if !certificate_path.try_exists()? {
-        write_new_https_cert(&certificate_path, private_path)?;
-    }
-
-    Ok(certificate_path)
-}
-
-/// Generate a new certificate for HTTPS and write it to the given path.
-///
-/// Use the given private key to sign the certificate.
-fn write_new_https_cert(certificate_path: &Path, private_path: &Path) -> Result<()> {
-    // Load the private key from the provided path
-    let private_key = {
-        let private_key_data = std::fs::read(private_path)?;
-        PKey::private_key_from_pem(&private_key_data)?
-    };
-
+fn build_x509(sockets: &[SocketAddr], private_key: PKey<Private>) -> Result<X509> {
     let name = {
         let mut name_builder = X509NameBuilder::new()?;
         name_builder.append_entry_by_text("CN", "localhost")?;
         name_builder.build()
     };
 
-    // Create a self-signed certificate
-    let certificate = {
-        let mut builder = X509::builder()?;
-        builder.set_version(2)?;
-        builder.set_subject_name(name.as_ref())?;
-        builder.set_issuer_name(name.as_ref())?;
-        builder.set_not_before(Asn1Time::days_from_now(0)?.as_ref())?;
-        builder.set_not_after(Asn1Time::days_from_now(365)?.as_ref())?;
-        builder.set_pubkey(private_key.as_ref())?;
-        builder.set_serial_number(
-            openssl::asn1::Asn1Integer::from_bn(openssl::bn::BigNum::from_u32(1)?.as_ref())?
-                .as_ref(),
-        )?;
-        builder.sign(private_key.as_ref(), MessageDigest::sha256())?;
-        builder.build()
-    };
-
-    // Write certificate to file
-    File::create_new(certificate_path)?.write_all(&certificate.to_pem()?)?;
+    let mut san_builder = SubjectAlternativeName::new();
+    san_builder.dns("localhost");
+    san_builder.dns("localhost.localdomain");
+    for socket in sockets {
+        san_builder.ip(socket.ip().to_string().as_str());
+    }
 
-    Ok(())
+    let mut builder = X509::builder()?;
+    builder.set_version(2)?;
+    builder.set_subject_name(name.as_ref())?;
+    builder.set_issuer_name(name.as_ref())?;
+    builder.append_extension(san_builder.build(&builder.x509v3_context(None, None))?)?;
+    builder.set_not_before(Asn1Time::days_from_now(0)?.as_ref())?;
+    builder.set_not_after(Asn1Time::days_from_now(365)?.as_ref())?;
+    builder.set_pubkey(private_key.as_ref())?;
+    builder.set_serial_number(
+        openssl::asn1::Asn1Integer::from_bn(openssl::bn::BigNum::from_u32(1)?.as_ref())?.as_ref(),
```

Comment:
This is just something copy-paste work and I just found it. The hardcode bn number 1 will cause potential collision. If we only use it for test, it's fine. Or, do we need to consider to generate a random bn here?